File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        def pathsum(node, sum = 0):
            if node and (not node.left) and (not node.right):
                return sum + node.val == targetSum

            left = pathsum(node.left,node.val+sum) if node.left else False
            right = pathsum(node.right,node.val+sum) if node.right else False
            return left or right
        return (root and pathsum(root)) or False

File: 0112-path-sum/0112-path-sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
    let found = 0;
    if(!root) return false;
    function sum(node, target){
        if(!node.left && !node.right){
            if(node.val == target) found++
            return
        }
        if(node.left) sum(node.left, target - node.val)
        if(node.right) sum(node.right, target - node.val)
    }

    sum(root,targetSum);
    return Boolean(found)
};